code review

yermasog opened the pull request.

yermasog committed:
maven project set up

yermasog committed:
updated readme with project plan

yermasog committed:
added data files

yermasog committed:
MVC and repo pattern structure set up

yermasog committed:
host model created

yermasog committed:
guest model created

yermasog committed:
reservation model created

yermasog committed:
files set up

yermasog committed:
main menu option enum created

yermasog committed:
created dataException

yermasog committed:
added guest and host to reservation model

yermasog committed:
set up resFileRepo

yermasog committed:
created readDateString to convert to localDate

yermasog committed:
created test data files, findAll() repo test passes

yermasog committed:
addReservation()

yermasog committed:
add() passes!

yermasog committed:
edit()

yermasog committed:
edit() test pass!

yermasog committed:
cancel() test passes!

yermasog committed:
refactored reservationFileRepo: moved guest and host to it's own file

yermasog committed:
findByEmail() service

yermasog committed:
created result class in domain layer

yermasog committed:
add() and validate() reservations

yermasog committed:
shouldAdd(), shouldNotAddNull() passes

yermasog committed:
validatedOverlap passes

yermasog committed:
editRes() passes

yermasog committed:
cancelRes() passes

yermasog committed:
controller, view, and app set up, we have a hello world!

yermasog committed:
menu option set up

yermasog committed:
viewResByHostEmail works!

yermasog committed:
cleaned up runMenuLoop and added enterToContinue()

yermasog committed:
refactored guest and host services

yermasog committed:
cancelRes() works

yermasog committed:
makeRes() works

yermasog committed:
editRes() works

yermasog committed:
cleaned up view

yermasog committed:
dates display in correct format

yermasog committed:
addRes() loop fixed

yermasog committed:
added loop to continue prompting if delete res doesn't exist

yermasog committed:
refactored code

yermasog committed:
added spring DI

oneowhat commented on src/main/java/learn/reservations/ui/View.java:
```diff
@@ -0,0 +1,106 @@
+            System.out.printf("%s. %s%n", option.getValue(), option.getMessage());
+        }
+        System.out.println("Select [0-4]");
+        int numberOption = Integer.parseInt(console.nextLine());
```
This crashes the app unless I provide an int, a `readInt` method that handles that error gracefully would be better.

oneowhat commented on src/main/java/learn/reservations/ui/Controller.java:
```diff
@@ -0,0 +1,262 @@
+        List<Reservation> futureReservations;
+        do {
+            reservations = service.findByEmail(hostEmail);
+            futureReservations = reservations.stream()
+                    .filter(r -> r.getStartDate().isAfter(LocalDate.now()))
+                    .sorted(Comparator.comparing(reservation -> reservation.getStartDate()))
+                    .collect(Collectors.toList());
```
Streams!

oneowhat reviewed (changes requested):
Grade: not yet

Overall good work, Yer. Your code is idiomatic, layers are well organized. 

Let's fix these items:
* Calculate overcharges by 1 day
* Bad number input crashes application

yermasog committed:
deleted readDate() and used localdate.parse() instead

yermasog committed:
fixed overcalculation of total by 1 day

yermasog committed:
added readInt() to controller

yermasog commented:
I fixed the over charge and added a readInt() method. I'm still working on some of the other pieces of feedback, but my app should meet all the requirements now.

oneowhat commented on src/main/java/learn/reservations/ui/Controller.java:
```diff
@@ -220,7 +216,7 @@ public void cancelRes() throws DataAccessException {
 
     private BigDecimal calculateTotal(Host host, LocalDate start, LocalDate end, BigDecimal total) {
-        for (LocalDate date = start; !date.isAfter(end); date = date.plusDays(1)) {
+        for (LocalDate date = start; date.isBefore(end); date = date.plusDays(1)) {
```
Gotcha! Bugs beware.. 🦟☠️

oneowhat commented on src/main/java/learn/reservations/ui/View.java:
```diff
@@ -20,14 +20,24 @@
         return MainMenuOption.fromValue(numberOption);
     }
 
+    private int readInt() {
+        while (true)
+            try {
+                return Integer.parseInt(console.nextLine());
+            } catch (NumberFormatException ex) {
+                System.out.println("Invalid number; please select [0-4]");
+            }
+
+    }
```
Looks great.

oneowhat commented on src/main/java/learn/reservations/ui/View.java:
```diff
@@ -66,11 +76,12 @@ public LocalDate promptDate(String message) throws DataAccessException {
+        try {
+            LocalDate date = LocalDate.parse(console.nextLine(), formatter);
+            return date;
+        } catch (Exception ex) {
```
Is there a more specific `Exception` that could be caught here?

oneowhat reviewed (approved):
Thanks for the updates, Yer. Much improved!

Some things to look out for in future projects: 
* You have some failing unit tests
* Test coverage is a little light (no tests for overlapping dates, calculate total)
* Catch specific exceptions
